code/modules/events/crystal_event.dm

```diff
@@ -0,0 +1,477 @@
+/datum/round_event/crystal_invasion/proc/spawn_portal(list/wave_type)
+	var/list/spawners = list()
+	for(var/obj/effect/landmark/event_spawn/temp in GLOB.generic_event_spawns)
+		if(QDELETED(temp))
+			continue
+		if(is_station_level(temp.loc.z))
+			spawners += temp
+
+	if(!spawners.len)
+		message_admins("No landmarks on the station, aborting")
+		return MAP_ERROR
+
+	var/pick_portal = pickweight(wave_type)
+	var/obj/spawner = pick_n_take(spawners)
+	new pick_portal(spawner.loc)
+
+/datum/round_event/crystal_invasion/proc/more_portals()
+	priority_announce("WARNING - Detected another spike from the destabilized crystal. More portals are spawning all around the station, the next spike could \
+						cause a \[REDACTED] class event we assume you have 10 more minutes before total crystal annihilation", "Alert")
+	sound_to_playing_players('sound/misc/notice1.ogg')
+	for(var/i = 0, i< 8, i++)
+		spawn_portal(GLOB.waves["small wave"])
+
+	for(var/C in GLOB.destabilized_crystals)
+		addtimer(CALLBACK(C, /obj/machinery/destabilized_crystal/proc/zk_event), 10 MINUTES)
+
+/*
+This section is for the destabilized SM
+*/
+/obj/machinery/destabilized_crystal
+	name = "destabilized crystal"
+	desc = "A strangely translucent and iridescent crystal."
+	icon = 'icons/obj/supermatter.dmi'
+	icon_state = "psy"
+	density = TRUE
+	anchored = TRUE
+	flags_1 = PREVENT_CONTENTS_EXPLOSION_1
+	light_range = 4
+	resistance_flags = INDESTRUCTIBLE | LAVA_PROOF | FIRE_PROOF | UNACIDABLE | ACID_PROOF | FREEZE_PROOF
+	///If not active the crystal will not emit radiations and gases
+	var/active = TRUE
+	///Check if the crew managed to stop the ZK-class event
+	var/is_zk = FALSE
+
+/obj/machinery/destabilized_crystal/Initialize()
+	. = ..()
+	GLOB.destabilized_crystals += src
+
+/obj/machinery/destabilized_crystal/Destroy()
+	GLOB.destabilized_crystals -= src
+	if(is_zk)
+		priority_announce("WARNING - Portal are appearing everywhere, you failed to contain the event. You people should feel ashamed of yourself!","Alarm")
+	return..()
+
+/obj/machinery/destabilized_crystal/process()
+	if(active)
+		if(prob(75))
+			src.fire_nuclear_particle()
+			radiation_pulse(src, 250, 6)
+		var/turf/T = loc
+		var/datum/gas_mixture/env = T.return_air()
+		var/datum/gas_mixture/removed
+		var/gasefficency = 0.5
+		removed = env.remove(gasefficency * env.total_moles())
+		removed.assert_gases(/datum/gas/bz, /datum/gas/miasma)
+		removed.gases[/datum/gas/bz][MOLES] += 15.5
+		removed.gases[/datum/gas/miasma][MOLES] += 5.5
+		env.merge(removed)
+		air_update_turf()
+
+/obj/machinery/destabilized_crystal/proc/zk_event()
+	active = FALSE
+	priority_announce("WARNING - The crystal has reached critical instability point. ZK-Event inbound, please do not panic, anyone who panics will \
+						be terminated on the spot. Have a nice day", "Alert")
+	sound_to_playing_players('sound/machines/alarm.ogg')
+	sleep(15 SECONDS)
```

Comment:
Replace this with a timer.

Reply:
can `addtimer` be used without a callback function? because i didn't want to make it into yet another proc

Reply:
It cannot, but it's fine to add another proc. Just add `do_zk_event()`